src/hotspot/share/gc/shared/classUnloadingContext.cpp

```diff
@@ -0,0 +1,174 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "classfile/classLoaderData.inline.hpp"
+#include "code/nmethod.hpp"
+#include "gc/shared/classUnloadingContext.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "utilities/growableArray.hpp"
+
+ClassUnloadingContext* ClassUnloadingContext::_context = nullptr;
+
+ClassUnloadingContext::ClassUnloadingContext(uint num_workers, bool lock_codeblob_free_separately) :
+  _cld_head(nullptr),
+  _num_nmethod_unlink_workers(num_workers),
+  _unlinked_nmethods(nullptr),
+  _lock_codeblob_free_separately(lock_codeblob_free_separately) {
+
+  assert(_context == nullptr, "context already set");
+  _context = this;
+
+  assert(num_workers > 0, "must be");
+
+  _unlinked_nmethods = NEW_C_HEAP_ARRAY(NMethodSet*, num_workers, mtGC);
+  for (uint i = 0; i < num_workers; ++i) {
+    _unlinked_nmethods[i] = new NMethodSet();
+  }
+}
+
+ClassUnloadingContext::~ClassUnloadingContext() {
+  for (uint i = 0; i < _num_nmethod_unlink_workers; ++i) {
+    delete _unlinked_nmethods[i];
+  }
+  FREE_C_HEAP_ARRAY(NMethodSet*, _unlinked_nmethods);
+
+  assert(_context == this, "context not set correctly");
+  _context = nullptr;
+}
+
+bool ClassUnloadingContext::has_unloaded_classes() const {
+  return _cld_head != nullptr;
+}
+
+void ClassUnloadingContext::register_unloading_class_loader_data(ClassLoaderData* cld) {
+  assert_locked_or_safepoint(ClassLoaderDataGraph_lock);
+
+  cld->unload();
+
+  cld->set_unloading_next(_cld_head);
+  _cld_head = cld;
+}
+
+void ClassUnloadingContext::purge_class_loader_data() {
+  for (ClassLoaderData* cld = _cld_head; cld != nullptr;) {
+    assert(cld->is_unloading(), "invariant");
+
+    ClassLoaderData* next = cld->unloading_next();
+    delete cld;
+    cld = next;
+  }
+}
+
+void ClassUnloadingContext::classes_unloading_do(void f(Klass* const)) {
+  assert_locked_or_safepoint(ClassLoaderDataGraph_lock);
+  for (ClassLoaderData* cld = _cld_head; cld != nullptr; cld = cld->unloading_next()) {
+    assert(cld->is_unloading(), "invariant");
+    cld->classes_do(f);
+  }
+}
```

Comment:
I don't understand why CLDG specific methods were moved here.  They should be unaware of nmethod purging. and these 4 methods don't have any nmethod purging in them either and are specific to the CLDG implementation.

Reply:
The idea is to have GC take control how the unloading CLDs are stored/which data structure it is going to use to manage them to ultimately allow more control about class unloading for parallelization.

Which on the one hand makes pauses shorter (for stw collectors), and on the other hand decreases the time the CLDG_lock is held (not sure it is nice that the concurrent collectors currently may hold that one for ~100ms in my test...).

I believe having the linked list of unloading CLDs embedded in the CLDs for use by the GC not only seems wrong (i.e. it's a GC data structure located in runtime code) but is also very limiting (need to have one for all, fixed singly linked list).

This change moves knowledge of how unloading CLDs are managed to GC area - runtime code just tells GC that a particular CLD is unloading.
(Currently the `ClassUnloadingContext` also calls the `unload` method during registration to keep current functionality, but the plan is to separate the step of registration and actual unloading to allow custom handling of the second part; the registering, although it's still walking a singly linked list, is comparatively fast).

These four methods provide a thin abstraction over the CLDs that are unloading (that runtime doesn't need and should not worry about imo).

With that in place it is possible to slice the actual unloading work into phases according to dependencies (depending on GC if desired), potentially overlapping with other existing phases in collectors already allowing that (e.g. the parallel code unloading, but that is only an implementation detail to reduce overall parallel phases), or even moving some of that work sometime else (the `CLD::unload()` method unfortunately currently may do some memory freeing too).

However most time is spent in notifying various components which can be parallelized (at least parallelize the different types of notifications).